### ISSUE TITLE
DeviceClient throws ProtocolException: Connection lost (32109) - java.io.EOFException

### ISSUE THREAD

**Issue by tianhai-liu** [Azure/azure-iot-sdk-java#570]
- **OS and version used:** <VERSION> Linux

- **Java runtime used:** <VERSION> 1.8

- **SDK version used:** <VERSION> 1.17.5

# Description of the issue:
The connection is lost and then the device client throws an exception `com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Mqtt connection lost`, and it connects again. Does this exception indicate that I don't use Azure API correctly? Or I can just ignore the exception? Thanks!

# Code sample exhibiting the issue:
```
while (true) {
    if (FileUploadStatusCallback is not invoked)  wait;
    File file = fetchNextFileToUpload();
    if (client == null) client = new DeviceClient(connectString, MQTT);
    client.open();
    client.setOperationTimeout(300000);
    client.uploadToBlobAsync(..., new IotHubEventCallback(...), file);
}
```

# Console log of the issue:

> FINE: Dev01: Stopping due to IOException 
> FINE: Dev01: state=DISCONNECTING 
> FINE: Dev01: stopping 
> FINE: Dev01: notify spaceAvailable 
> FINE: Dev01: notify workAvailable and wait for run 
> FINE: Dev01: stopped 
> FINE: Dev01: stopping 
> FINE: Dev01: stopped 
> FINE: Dev01: resp=Client is currently disconnecting (32102) 
> FINE: Dev01: > 
> FINE: Dev01: key=Ping token=org.eclipse.paho.client.mqttv3.MqttToken@38d2f8b2 
> FINE: Dev01: reason Connection lost (32109) - java.io.EOFException 
> FINE: Dev01: > 
> FINE: Dev01: key=Ping 
> FINE: Dev01: callback and notify for key=Ping 
> FINE: Dev01: received key=0 token=org.eclipse.paho.client.mqttv3.MqttToken@38d2f8b2 message=PINGRESP msgId 0 
> FINE: Dev01: >key=Ping response=PINGRESP msgId 0 excep=null 
> FINE: Dev01: disconnected 
> FINE: Dev01: stopping sender 
> FINE: Dev01: no outstanding flows and not connected 
> FINE: Dev01: stopped 
> FINE: Dev01: get message returned null, stopping} 
> FINE: null: stop 
> FINE: Dev01: < 
> FINE: Dev01: state=DISCONNECTED 
> FINE: Dev01: call connectionLost 
> FINE: Dev01: < 
> FINE: Dev01: > 0 tokens 
> FINE: Dev01: > 
> FINE: Dev01: < 
> FINE: IotHubConnectionStatusChangeCallback: DISCONNECTED_RETRYING
> com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Mqtt connection lost
> 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException(PahoExceptionTranslator.java:75)
> 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.connectionLost(Mqtt.java:366)
> 	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.connectionLost(CommsCallback.java:292)
> 	at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:423)
> 	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:181)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:660)
> 	at java.lang.Thread.run(Thread.java:1652)
> Caused by: Connection lost (32109) - java.io.EOFException
> 	... 12 more
> Caused by: java.io.EOFException
> 	at java.io.DataInputStream.readByte(DataInputStream.java)
> 	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:92)
> 	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:133)
> 	... 11 more
>  
> FINE: Dev01: <init> ClientID=Dev01 ServerURI=ssl://<host>:8883 PersistenceType=org.eclipse.paho.client.mqttv3.persist.MemoryPersistence@214aa04a 
> FINE: Dev01: <> 
> FINER: Dev01:  
> FINE: Dev01: > 
> FINE: Dev01: cleanSession=false connectionTimeout=30 TimekeepAlive=230 userName=<host>/Dev01/?api-version=2018-06-30&DeviceClientType=com.microsoft.azure.sdk.iot.iot-device-client%2F1.17.5%20%281.8.0_131%3B%20Linux%3B%20arm%29 password=[notnull] will=[null] userContext=null callback=null 
> FINE: Dev01: URI=ssl://<host>:8883 
> FINE: Dev01: URI=ssl://<host>:8883 
> FINE: Dev01: < 
> FINE: Dev01: state=CONNECTING 
> FINE: Dev01: > 
> FINE: Dev01: > 
> FINE: Dev01: > 
> FINE: Dev01: key=null wait max=60,000 token=key=null ,topics= ,usercontext=org.eclipse.paho.client.mqttv3.MqttAsyncClient@214ae00a ,isComplete=false ,isNotified=false ,exception=null ,actioncallback=org.eclipse.paho.client.mqttv3.internal.ConnectActionListener@741f50bf 
> FINE: Dev01: key=Con message=CONNECT clientId Dev01 keepAliveInterval 230 
> FINE: Dev01: >key=null timeout=60,000 sent=false completed=false hasException=false response=null token=key=null ,topics= ,usercontext=org.eclipse.paho.client.mqttv3.MqttAsyncClient@214ae00a ,isComplete=false ,isNotified=false ,exception=null ,actioncallback=org.eclipse.paho.client.mqttv3.internal.ConnectActionListener@741f50bf 
> FINE: Dev01: key=Con token=org.eclipse.paho.client.mqttv3.MqttToken@741f907f 
> FINE: Dev01: key=null wait max=60,000 
> FINE: Dev01: connect to host <host> port 8,883 timeout 30,000 
> FINE: Dev01: starting 
> FINE: Dev01: network read message 
> FINE: Dev01: wait for new work or for space in the inflight window 
> FINE: Dev01: wait for workAvailable 
> FINE: Dev01: 200 
> FINE: Dev01: pending send key=0 message CONNECT clientId Dev01 keepAliveInterval 230 
> FINE: Dev01: key=Con message=CONNECT clientId Dev01 keepAliveInterval 230 
> FINE: Dev01: key=Con token=org.eclipse.paho.client.mqttv3.MqttToken@741f907f 
> FINE: Dev01: new work or ping arrived  
> FINE: Dev01: network send key=Con msg=CONNECT clientId Dev01 keepAliveInterval 230 
> FINE: Dev01: sent bytes count=13 
> FINE: Dev01: sent bytes count=321 
> FINE: null: Sent CONNECT clientId Dev01 keepAliveInterval 230 
> FINE: Dev01: key=Con 
> FINE: Dev01: > key=Con 
> FINE: Dev01: wait for new work or for space in the inflight window 
> FINE: Dev01: received bytes count=1 
> FINE: Dev01: received bytes count=2 
> FINE: null: Automatic Reconnect Successful: CONNACK msgId 0 session present:true return code: 0 
> FINE: Dev01: received key=0 message=CONNACK msgId 0 session present:true return code: 0 
> FINE: Dev01: connected 
> FINE: null: start timer for client:Dev01 
> FINE: Dev01: state=CONNECTED 
> FINE: Dev01: >key=Con response=CONNACK msgId 0 session present:true return code: 0 excep=null 
> FINE: Dev01: >key=Con response=CONNACK msgId 0 session present:true return code: 0 excep=null 
> FINE: Dev01: keyCon, msg=CONNACK msgId 0 session present:true return code: 0, excep=null 
> FINE: Dev01: new workAvailable. key=Con 
> FINE: Dev01: key=Con 
> FINE: Dev01: network read message 
> FINE: Dev01: new work or ping arrived  
> FINE: Dev01: callback and notify for key=Con 
> FINE: Dev01: wait for new work or for space in the inflight window 
> FINE: Dev01: received key=0 token=org.eclipse.paho.client.mqttv3.MqttToken@741f907f message=CONNACK msgId 0 session present:true return code: 0 
> FINE: Dev01: >key=Con response=CONNACK msgId 0 session present:true return code: 0 excep=null 
> FINE: Dev01: call onSuccess key=Con 
> FINE: Dev01: >key=null response=CONNACK msgId 0 session present:true return code: 0 excep=null 
> FINE: Dev01: >key=null response=CONNACK msgId 0 session present:true return code: 0 excep=null 
> FINE: Dev01: notify spaceAvailable 
> FINE: Dev01: wait for workAvailable 
> FINE: Dev01: key=null response=CONNACK msgId 0 session present:true return code: 0 
> FINE: Dev01: Subscribe topicFilter=topic=devices/Dev01/messages/devicebound/# qos=1 userContext=null callback=null 
> FINE: Dev01: 200 
> FINE: Dev01: pending send key=1 message SUBSCRIBE names:["devices/Dev01/messages/devicebound/#"] qos:[1] 
> FINE: Dev01: key=1 message=SUBSCRIBE names:["devices/Dev01/messages/devicebound/#"] qos:[1] 
> FINE: Dev01: key=1 token=org.eclipse.paho.client.mqttv3.MqttToken@2843c703 
> FINE: Dev01: < 
> FINE: Dev01: new work or ping arrived  
> FINE: Dev01: key=1 wait max=15,000 token=key=1 ,topics=devices/Dev01/messages/devicebound/#,  ,usercontext=null ,isComplete=false ,isNotified=false ,exception=null ,actioncallback=null 
> FINE: Dev01: network send key=1 msg=SUBSCRIBE names:["devices/Dev01/messages/devicebound/#"] qos:[1] 
> FINE: Dev01: >key=1 timeout=15,000 sent=false completed=false hasException=false response=null token=key=1 ,topics=devices/Dev01/messages/devicebound/#,  ,usercontext=null ,isComplete=false ,isNotified=false ,exception=null ,actioncallback=null 
> FINE: Dev01: sent bytes count=4 
> FINE: Dev01: key=1 wait max=15,000 
> FINE: Dev01: sent bytes count=39 
> FINE: null: Sent SUBSCRIBE names:["devices/Dev01/messages/devicebound/#"] qos:[1] 
> FINE: Dev01: key=1 
> FINE: Dev01: > key=1 
> FINE: Dev01: wait for new work or for space in the inflight window 
> FINE: Dev01: received bytes count=1 
> FINE: Dev01: received bytes count=3 
> FINE: null: Automatic Reconnect Successful: SUBACK msgId 1 granted Qos 1 
> FINE: Dev01: received key=1 message=SUBACK msgId 1 granted Qos 1 
> FINE: Dev01: >key=1 response=SUBACK msgId 1 granted Qos 1 excep=null 
> FINE: Dev01: >key=1 response=SUBACK msgId 1 granted Qos 1 excep=null 
> FINE: Dev01: key1, msg=SUBACK msgId 1 granted Qos 1, excep=null 
> FINE: Dev01: key=1 response=SUBACK msgId 1 granted Qos 1 
> FINE: Dev01: new workAvailable. key=1 
> FINE: IotHubConnectionStatusChangeCallback: CONNECTED

**Comment by timtay-microsoft** [comment]
The EOFException is most commonly seen when two different device clients try to open a connection to IotHub under the same device identity at the same time like the below sample:

```java
String connectionString = "<a device connection string>";
DeviceClient client1 = new DeviceClient(connectionString, MQTT);
DeviceClient client2 = new DeviceClient(connectionString, MQTT);
client1.open();
client2.open();
//EOFException should be thrown by each client
```

I recommend not creating multiple device client instances under the same identity, but instead re-using just one client for all operations for a given device identity

**Comment by timtay-microsoft** [comment]
@tianhai-liu are you still seeing any issues?

**Comment by tianhai-aicas** [comment]
@timtay-microsoft thanks. I have not more issues in this thread.

**Comment by timtay-microsoft** [comment]
Okay, I'm going to go ahead and close this thread then. Feel free to open a new issue if you have any other questions

**Comment by az-iot-builder-01** [comment]
@tianhai-liu, @tianhai-aicas, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by tianhai-liu** [comment]
I have the same issue happens again in version 1.17.5. Are the fixes in the latest version 1.18.0?
